dmoj/judge.py

```diff
@@ -284,22 +226,235 @@ def listen(self):
+        self.abort_grading()
         self.updater_exit = True
         self.updater_signal.set()
         if self.packet_manager:
             self.packet_manager.close()
 
+    def log_internal_error(self, exc=None, message=None):
+        if not message:
+            # If exc is exists, raise it so that sys.exc_info() is populated with its data
+            if exc:
+                try:
+                    raise exc
+                except KeyboardInterrupt:
+                    # Let KeyboardInterrupt bubble up.
+                    raise
+                except:  # noqa: E722, we want to catch everything
+                    pass
+
+            exc = sys.exc_info()
+            message = ''.join(traceback.format_exception(*exc))
+
+        # Strip ANSI from the message, since this might be a checker's CompileError
+        # ...we don't want to see the raw ANSI codes from GCC/Clang on the site.
+        # We could use format_ansi and send HTML to the site, but the site doesn't presently support HTML
+        # internal error formatting.
+        self.packet_manager.internal_error_packet(strip_ansi(message))
+
+        # Logs can contain ANSI, and it'll display fine
+        print(message, file=sys.stderr)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        pass
+
+
+class SubmissionJudgeProcess:
+    def __init__(self, submission_id, problem_id, language, source, time_limit, memory_limit, short_circuit, meta):
+        self.submission_id = submission_id
+        self.problem_id = problem_id
+        self.language = language
+        self.source = source
+        self.time_limit = time_limit
+        self.memory_limit = memory_limit
+        self.short_circuit = short_circuit
+        self.meta = meta
+
+        self._abort_requested = False
+        self.grader = None
+        self.process = None
+        self.process_conn = None
+
+    def run(self):
+        self.process_conn, child_conn = multiprocessing.Pipe()
+        self.process = multiprocessing.Process(
+            name="DMOJ Judge Handler for %s/%d" % (self.problem_id, self.submission_id),
+            target=self._run_grading_wrapper_process,
+            args=(child_conn,),
+        )
+        self.process.start()
+        child_conn.close()
+
+        while True:
+            try:
+                ipc_msg = self.process_conn.recv()
+            except Exception:
+                logger.error("Failed to read IPC message from handler!")
+                raise
+
+            if ipc_msg is None:
+                self.process_conn.send(None)
+                break
+
+            yield ipc_msg
+
+        self.process_conn.close()
+        self.process.join()
+
+    def stop(self):
+        if self.process:
+            # Might be None if run was never called, or failed.
+            try:
+                self.process.join(timeout=IPC_TEARDOWN_TIMEOUT)
+                if self.process.is_alive():
```

Comment:
Maybe this should go into finally... In case something bad happens, we still want to kill the process.

Reply:
Good idea.